src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java

```diff
@@ -8,22 +15,73 @@
 import org.springframework.context.annotation.Profile;
 import org.springframework.stereotype.Component;
 
-import java.util.List;
-
 @Component
 @Profile("!test")
 public class KafkaFactory {
+
     private final KafkaLocationManager kafkaLocationManager;
-    private final KafkaProducer<String, String> kafkaProducer;
+    @Nullable
+    private Producer<String, String> activeProducer;
+    private final Map<Producer<String, String>, AtomicInteger> useCount = new ConcurrentHashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
 
     @Autowired
     public KafkaFactory(final KafkaLocationManager kafkaLocationManager) {
         this.kafkaLocationManager = kafkaLocationManager;
-        kafkaProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
     }
 
-    public Producer<String, String> getProducer() {
-        return kafkaProducer;
+
+    public Producer<String, String> takeProducer() {
+        rwLock.readLock().lock();
+        try {
+            if (null != activeProducer) {
+                useCount.get(activeProducer).incrementAndGet();
+                return activeProducer;
+            }
+        } finally {
+            rwLock.readLock().unlock();
+        }
+        rwLock.writeLock().lock();
+        try {
+            if (null != activeProducer) {
+                useCount.get(activeProducer).incrementAndGet();
+                return activeProducer;
+            }
+            activeProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
+            useCount.put(activeProducer, new AtomicInteger(1));
+            return activeProducer;
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    public void releaseProducer(final Producer<String, String> producer) {
+        final int newUseCount = useCount.get(producer).decrementAndGet();
```

Comment:
And fixed implementation: https://github.com/zalando/nakadi/pull/443/commits/f793a564f0541cba936210948d5d2707cb1e474b